File: hw9/memc_load.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import Queue
import os
import gzip
import sys
import glob
import time
import logging
import collections
import threading
import multiprocessing
from itertools import islice
from optparse import OptionParser
# brew install protobuf
# protoc  --python_out=. ./appsinstalled.proto
# pip install protobuf
import appsinstalled_pb2
# pip install python-memcached
import memcache

MEMCACHE_SOCKET_TIMEOUT = 2
NORMAL_ERR_RATE = 0.01
CHUNK_SIZE = 100
SENTINEL = object()

AppsInstalled = collections.namedtuple('AppsInstalled', ['dev_type', 'dev_id', 'lat', 'lon', 'apps'])


class MemcachedThread(threading.Thread):
    def __init__(self, memc_addr, queue, stats_queue, dry_run):
        threading.Thread.__init__(self)
        self.daemon = True
        self.memc_addr = memc_addr
        self.queue = queue
        self.stats_queue = stats_queue
        self.dry_run = dry_run
        self.processed = 0
        self.errors = 0

    @staticmethod
    def _set_multi(connection, keys_map, tries=3, delay=0.5, backoff=2):
        failed_keys = connection.set_multi(keys_map)
        mtries, mdelay = tries, delay
        while failed_keys and mtries > 0:
            time.sleep(mdelay)
            keys_map = {k: v for k, v in keys_map.items() if k in failed_keys}
            failed_keys = connection.set_multi(keys_map)
            mtries -= 1
            mdelay *= backoff
        return failed_keys

    def run(self):
        connection = memcache.Client([self.memc_addr], socket_timeout=MEMCACHE_SOCKET_TIMEOUT)
        while True:
            try:
                tasks = self.queue.get(timeout=0.1)

                if tasks == SENTINEL:
                    self.stats_queue.put((self.processed, self.errors))
                    logging.info('%s | %s | Records processed: %d | Records errors: %d',
                                 multiprocessing.current_process().name,
                                 threading.current_thread().name, self.processed, self.errors)
                    self.queue.task_done()
                    break

                try:
                    if self.dry_run:
                        for task in tasks:
                            key, ua = task
                            logging.debug('%s - %s -> %s', self.memc_addr, key, str(ua).replace('\n', ' '))
                    else:
                        keys_map = {k: v.SerializeToString() for k, v in tasks}
                        failed_keys = self._set_multi(connection, keys_map)
                        self.processed += len(tasks) - len(failed_keys)
                        self.errors += len(failed_keys)
                except Exception as e:
                    logging.exception('Cannot write to memc %s: %s', self.memc_addr, e)

                self.queue.task_done()
            except Queue.Empty:
                continue


def dot_rename(path):
    head, fn = os.path.split(path)
    # atomic in most cases
    os.rename(path, os.path.join(head, '.' + fn))


def buf_appsinstalled(appsinstalled):
    ua = appsinstalled_pb2.UserApps()
    ua.lat = appsinstalled.lat
    ua.lon = appsinstalled.lon
    key = '{}:{}'.format(appsinstalled.dev_type, appsinstalled.dev_id)
    ua.apps.extend(appsinstalled.apps)
    return key, ua


def parse_appsinstalled(line):
    line_parts = line.strip().split('\t')
    if len(line_parts) < 5:
        return
    dev_type, dev_id, lat, lon, raw_apps = line_parts
    if not dev_type or not dev_id:
        return
    try:
        apps = [int(a.strip()) for a in raw_apps.split(',')]
    except ValueError:
        apps = [int(a.strip()) for a in raw_apps.split(',') if a.isidigit()]
        logging.info('Not all user apps are digits: `%s`', line)
    try:
        lat, lon = float(lat), float(lon)
    except ValueError:
        logging.info('Invalid geo coords: `%s`', line)
    return AppsInstalled(dev_type, dev_id, lat, lon, apps)


def lines_chunk_handler(chunk, conn_pools):
    errors = 0

    packed_chunks = collections.defaultdict(list)

    for line in chunk:
        line = line.strip()

        if not line:
            continue

        appsinstalled = parse_appsinstalled(line)

        if not appsinstalled:
            errors += 1
            continue

        key, ua = buf_appsinstalled(appsinstalled)

        packed_chunks[appsinstalled.dev_type].append((key, ua))

    for k, v in packed_chunks.items():
        conn_queue = conn_pools.get(k)
        if not conn_queue:
            errors += 1
            logging.error('Unknow device type: %s', k)
            continue
        conn_queue.put(v)

    return errors


def handle_log((fn, device_memc, dry_run)):
    stats_queue = Queue.Queue()
    connections = []
    conn_pools = collections.defaultdict(Queue.Queue)

    # Start memcached connection handle threads
    for k, addr in device_memc.items():
        memcached_thread = MemcachedThread(addr, conn_pools[k], stats_queue, dry_run)
        connections.append(memcached_thread)

    for memcached_thread in connections:
        memcached_thread.start()

    logging.info('%s | Processing %s', multiprocessing.current_process().name, fn)

    processed = read = errors = 0
    with gzip.open(fn) as fd:
        chunk = list(islice(fd, CHUNK_SIZE))
        while all(c.is_alive() for c in connections):
            if not chunk:
                break
            errors += lines_chunk_handler(chunk, conn_pools)
            read += len(chunk)
            chunk = list(islice(fd, CHUNK_SIZE))

    logging.info('%s | File: %s | Total records readed: %s', multiprocessing.current_process().name, fn,
                 read)

    # Finish memcached connection handle threads
    for k, addr in device_memc.items():
        conn_pools[k].put(SENTINEL)

    for memcached_thread in connections:
        memcached_thread.join()

    while not stats_queue.empty():
        worker_processed, worker_errors = stats_queue.get(timeout=0.1)
        processed += worker_processed
        errors += worker_errors

    logging.info('%s | File: %s | Total records processed: %d | Total records errors: %d',
                 multiprocessing.current_process().name, fn, processed, errors)

    if processed:
        err_rate = float(errors) / processed
        if err_rate < NORMAL_ERR_RATE:
            logging.info('%s | File: %s | Acceptable error rate (%s). Successfull load',
                         multiprocessing.current_process().name, fn, err_rate)
        else:
            logging.error('%s | File: %s | High error rate (%s > %s). Failed load',
                          multiprocessing.current_process().name, fn, err_rate,
                          NORMAL_ERR_RATE)

    return fn


def main(options):
    device_memc = {
        'idfa': options.idfa,
        'gaid': options.gaid,
        'adid': options.adid,
        'dvid': options.dvid,
    }

    pool = multiprocessing.Pool(int(options.workers))
    fargs = []

    for fn in glob.iglob(options.pattern):
        fargs.append((fn, device_memc, options.dry))

    fargs.sort(key=lambda x: x[0])

    for f in pool.imap(handle_log, fargs):
        dot_rename(f)
        logging.info('%s | Renamed file %s', multiprocessing.current_process().name, f)


def prototest():
    sample = 'idfa\t1rfw452y52g2gq4g\t55.55\t42.42\t1423,43,567,3,7,23\ngaid\t7rfw452y52g2gq4g\t55.55\t42.42\t7423,424'
    for line in sample.splitlines():
        dev_type, dev_id, lat, lon, raw_apps = line.strip().split('\t')
        apps = [int(a) for a in raw_apps.split(',') if a.isdigit()]
        lat, lon = float(lat), float(lon)
        ua = appsinstalled_pb2.UserApps()
        ua.lat = lat
        ua.lon = lon
        ua.apps.extend(apps)
        packed = ua.SerializeToString()
        unpacked = appsinstalled_pb2.UserApps()
        unpacked.ParseFromString(packed)
        assert ua == unpacked


if __name__ == '__main__':
    op = OptionParser()
    op.add_option('-t', '--test', action='store_true', default=False)
    op.add_option('-l', '--log', action='store', default=None)
    op.add_option('--dry', action='store_true', default=False)
    op.add_option('--pattern', action='store', default='/data/appsinstalled/*.tsv.gz')
    op.add_option('--idfa', action='store', default='127.0.0.1:33013')
    op.add_option('--gaid', action='store', default='127.0.0.1:33014')
    op.add_option('--adid', action='store', default='127.0.0.1:33015')
    op.add_option('--dvid', action='store', default='127.0.0.1:33016')
    op.add_option('--workers', action='store', default=2)
    (opts, args) = op.parse_args()
    logging.basicConfig(filename=opts.log, level=logging.INFO if not opts.dry else logging.DEBUG,
                        format='[%(asctime)s] %(levelname).1s %(message)s', datefmt='%Y.%m.%d %H:%M:%S')
    if opts.test:
        prototest()
        sys.exit(0)

    logging.info('Memc loader started with options: %s', opts)

    try:
        main(opts)
    except Exception as e:
        logging.exception('Unexpected error: %s', e)
        sys.exit(1)

File: hw11/memc_load.go
package main

import (
	"bufio"
	"compress/gzip"
	"errors"
	"fmt"
	"github.com/AndrewLrrr/memclog/appsinstalled"
	"github.com/bradfitz/gomemcache/memcache"
	"github.com/golang/protobuf/proto"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"time"
)

const (
	LineWorkers     = 100
	ChannelsBuffer  = 100
	LogsDir         = "./logs"
	MemcacheTimeout = 500 * time.Millisecond
	NormalErrRate   = 0.01
)

type LogLine struct {
	devType, devId string
	lat, lon       float64
	apps           []uint32
}

type MemcacheTask struct {
	key   string
	value []byte
}

type Statistic struct {
	processed int
	errors    int
}

type Memcache struct {
	connection *memcache.Client
}

func NewMemcache(addr string) *Memcache {
	mc := Memcache{}
	mc.connection = memcache.New(addr)
	mc.connection.Timeout = MemcacheTimeout
	return &mc
}

func (mc *Memcache) setItem(key string, value []byte) error {
	tries := 3
	delay := 100
	backoff := 2
	if err := mc.connection.Set(&memcache.Item{Key: key, Value: value}); err != nil {
		for {
			time.Sleep(time.Duration(delay) * time.Millisecond)
			if err != nil && tries > 0 {
				err = mc.connection.Set(&memcache.Item{Key: key, Value: value})
				tries--
				delay *= backoff
			} else {
				return err
			}
		}
	}
	return nil
}

func lineWorker(channels map[string](chan *MemcacheTask), queue <-chan string, stats chan *Statistic) {
	processErrors := 0
	for {
		line, ok := <-queue
		if !ok {
			stat := &Statistic{
				processed: 0,
				errors:    processErrors,
			}
			stats <- stat
			return
		}
		logLine, err := parseLine(line)
		if err != nil {
			processErrors++
			log.Println(err)
		} else {
			packed, err := protobufLine(logLine)
			if err != nil {
				processErrors++
				log.Println(err)
			} else {
				key := fmt.Sprintf("%s:%s", logLine.devType, logLine.devId)
				task := &MemcacheTask{
					key:   key,
					value: packed,
				}
				channels[logLine.devType] <- task
			}
		}
	}
}

func memcacheWorker(mc *Memcache, queue <-chan *MemcacheTask, stats chan *Statistic) {
	processSuccess := 0
	processErrors := 0
	for {
		task, ok := <-queue
		if !ok {
			stat := &Statistic{
				processed: processSuccess,
				errors:    processErrors,
			}
			stats <- stat
			log.Printf("memcache processed %d | errors %d", processSuccess, processErrors)
			return
		}
		err := mc.setItem(task.key, task.value)
		if err != nil {
			processErrors++
			log.Println(err)
		} else {
			processSuccess++
		}
	}
}

func dotRename(path string) error {
	if name := filepath.Base(path); !strings.HasPrefix(name, ".") {
		newName := "." + name
		dir := filepath.Dir(path)
		newPath := filepath.Join(dir, newName)
		return os.Rename(path, newPath)
	}
	return nil
}

func parseLine(line string) (LogLine, error) {
	lineParts := strings.Split(strings.TrimSpace(line), "\t")
	logLine := LogLine{}

	if len(lineParts) != 5 {
		return logLine, errors.New("incorrect log line")
	}

	logLine.devType = strings.TrimSpace(lineParts[0])
	logLine.devId = strings.TrimSpace(lineParts[1])

	if len(logLine.devId) == 0 {
		return logLine, errors.New("empty dev_id is not allowed")
	}

	if len(logLine.devType) == 0 {
		return logLine, errors.New("empty dev_type is not allowed")
	}

	if lat, err := strconv.ParseFloat(lineParts[2], 64); err == nil {
		logLine.lat = lat
	} else {
		logLine.lat = 0
		log.Printf("invalid latitude: `%s`", line)
	}

	if lon, err := strconv.ParseFloat(lineParts[3], 64); err == nil {
		logLine.lon = lon
	} else {
		logLine.lon = 0
		log.Printf("invalid longitude: `%s`", line)
	}

	apps := strings.Split(strings.TrimSpace(lineParts[4]), ",")

	for _, app := range apps {
		if num, err := strconv.ParseUint(app, 10, 32); err == nil {
			logLine.apps = append(logLine.apps, uint32(num))
		} else {
			log.Printf("not all user apps are digits: `%s`", line)
		}
	}

	return logLine, nil
}

func protobufLine(logLine LogLine) ([]byte, error) {
	packed := appsinstalled.UserApps{
		Apps: logLine.apps,
		Lat:  &logLine.lat,
		Lon:  &logLine.lon,
	}

	data, err := proto.Marshal(&packed)
	if err != nil {
		return nil, errors.New(fmt.Sprintf("marshaling error: %s", err))
	}

	return data, nil
}

func fileHandler(filePath string, connections map[string]*Memcache, statistic chan *Statistic) (map[string]int, error) {
	channels := make(map[string](chan *MemcacheTask))
	for key, connection := range connections {
		channels[key] = make(chan *MemcacheTask, ChannelsBuffer)
		go memcacheWorker(connection, channels[key], statistic)
	}

	read := 0
	lines := make(chan string, ChannelsBuffer)

	for i := 0; i < LineWorkers; i++ {
		go lineWorker(channels, lines, statistic)
	}

	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}

	gz, err := gzip.NewReader(file)
	if err != nil {
		return nil, err
	}

	scanner := bufio.NewScanner(gz)

	for scanner.Scan() {
		lines <- scanner.Text()
		read += 1
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	defer gz.Close()
	defer file.Close()

	close(lines)

	result := map[string]int{
		"read": read,
		"processSuccess": 0,
		"processErrors": 0,
	}

	for j := 0; j < LineWorkers; j++ {
		stat := <-statistic
		result["processSuccess"] += stat.processed
		result["processErrors"] += stat.errors
	}

	for _, channel := range channels {
		close(channel)
		stat := <-statistic
		result["processSuccess"] += stat.processed
		result["processErrors"] += stat.errors
	}

	return result, nil
}

func main() {
	num := runtime.NumCPU()
	runtime.GOMAXPROCS(num)

	files, err := ioutil.ReadDir(LogsDir)
	if err != nil {
		log.Fatal(err)
	}

	connections := map[string]*Memcache{
		"idfa": NewMemcache("127.0.0.1:33013"),
		"gaid": NewMemcache("127.0.0.1:33014"),
		"adid": NewMemcache("127.0.0.1:33015"),
		"dvid": NewMemcache("127.0.0.1:33016"),
	}

	statistic := make(chan *Statistic)
	filePaths := []string{}

	for _, file := range files {
		r, err := regexp.MatchString("^[^\\.]+\\.tsv\\.gz$", file.Name())
		if err == nil && r {
			filePaths = append(filePaths, filepath.Join(LogsDir, file.Name()))
		}
	}

	sort.Strings(filePaths)

	totalRead := 0
	totalProcessed := 0
	totalErrors := 0
	for _, filePath := range filePaths {
		log.Printf("start handle file %s", filePath)

		result, err := fileHandler(filePath, connections, statistic)

		if err != nil {
			log.Fatal(err)
		}

		dotRename(filePath)
		log.Printf("finish handle file %s", filePath)

		totalRead += result["read"]
		totalProcessed += result["processSuccess"]
		totalErrors += result["processErrors"]
		log.Printf("read lines: %d", result["read"])
		log.Printf("processed lines: %d", result["processSuccess"])
		log.Printf("error lines: %d", result["processErrors"])

		if result["processSuccess"] > 0 {
			errRate := float64(result["processErrors"] / result["processSuccess"])
			if errRate < NormalErrRate {
				log.Printf("file: %s | Acceptable error rate (%f). Successfull load", filePath, errRate)
			} else {
				log.Printf("file: %s | High error rate (%f > %f). Failed load", filePath, errRate, NormalErrRate)
			}
		}
		log.Println("----------------------------------------------")
	}

	close(statistic)

	log.Printf("total read lines: %d", totalRead)
	log.Printf("total processed lines: %d", totalProcessed)
	log.Printf("total error lines: %d", totalErrors)
}
